File: simulators/de.unistuttgart.iste.sqa.mpw.framework.java/core/src/main/java/de/unistuttgart/iste/sqa/mpw/framework/viewmodel/impl/GameViewPresenterBase.java
package de.unistuttgart.iste.sqa.mpw.framework.viewmodel.impl;

import de.unistuttgart.iste.sqa.mpw.framework.utils.Preconditions;
import de.unistuttgart.iste.sqa.mpw.framework.mpw.*;
import de.unistuttgart.iste.sqa.mpw.framework.datatypes.*;
import de.unistuttgart.iste.sqa.mpw.framework.facade.MiniProgrammingWorld;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.GameViewPresenter;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCell;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelLogEntry;
import javafx.beans.property.ReadOnlyListProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Semaphore;

public abstract class GameViewPresenterBase extends GameViewPresenter {
	private final MiniProgrammingWorld miniProgrammingWorld;

	private final Map<LogEntry, ViewModelLogEntry> logEntryMap = new HashMap<>();
	private final Map<Tile, ListChangeListener<TileContent>> tileContentsChangeListeners = new HashMap<>();

	public GameViewPresenterBase(final MiniProgrammingWorld miniProgrammingWorld) {
		super(new Semaphore(1, true));
		this.miniProgrammingWorld = miniProgrammingWorld;
	}

	@Override
	public void bind() {
		runLocked(() -> {
			bindSize();
			bindTiles();
			bindGameLog();
			bindButtons();
			onBind();
		});
	}

	// needs to be implemented by derived presenter classes, since here
	// no concrete Stage is available.
	protected abstract ReadOnlyObjectProperty<Size> getStageSizeFromConcreteStage();

	protected void onBind() {
		// can be overridden by subclasses
	}

	private void bindSize() {
		final ReadOnlyObjectProperty<Size> sizeProperty = getStageSizeFromConcreteStage();
		getViewModel().init(sizeProperty.get());
		sizeProperty.addListener((observableValue, oldSize, newSize) -> {
			runLocked(() -> {
				getViewModel().init(newSize);
			});
		});
	}

	private void bindTiles() {
		final var tilesProperty = getTilesPropertyFromConcreteStage();
		tilesProperty.addListener(tilesChangedListener);
		tilesProperty.forEach(this::addTileNode);
	}

	// needs to be implemented by derived presenter classes, since here
	// no concrete Stage is available.
	protected abstract ReadOnlyListProperty<Tile> getTilesPropertyFromConcreteStage();

	private void bindGameLog() {
		final var gameLog = miniProgrammingWorld.getGameLog();
		final var logEntriesProperty = gameLog.logEntriesProperty();
		logEntriesProperty.addListener(logChangedListener);
		logEntriesProperty.forEach(this::addLogEntry);
	}

	private void bindButtons() {
		final var viewModel = getViewModel();
		final var modeProperty = miniProgrammingWorld.getPerformance().modeProperty();
		final var gameCommandStack = miniProgrammingWorld.getPerformance().getGameCommandStack();
		final var anyExecutedCommandsBinding = gameCommandStack.executedCommandsProperty().emptyProperty().not();
		final var anyUndoneCommandsBinding = gameCommandStack.undoneCommandsProperty().emptyProperty().not();

		viewModel.playButtonEnabledProperty().bind(modeProperty.isEqualTo(Mode.PAUSED));
		viewModel.pauseButtonEnabledProperty().bind(modeProperty.isEqualTo(Mode.RUNNING));
		viewModel.undoButtonEnabledProperty().bind(anyExecutedCommandsBinding.and(modeProperty.isNotEqualTo(Mode.RUNNING)));
		viewModel.redoButtonEnabledProperty().bind(anyUndoneCommandsBinding.and(modeProperty.isNotEqualTo(Mode.RUNNING)));
	}

	/**
	 * Log Handling
	 */

	private final ListChangeListener<LogEntry> logChangedListener = new ListChangeListener<>() {
		@Override
		public void onChanged(final Change<? extends LogEntry> change) {
			runLocked(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(entry -> addLogEntry(entry));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(entry -> removeLogEntry(entry));
					}
				}
			});
		}
	};

	private void addLogEntry(final LogEntry entry) {
		final var viewModelLogEntry = new ViewModelLogEntry();
		viewModelLogEntry.setMessage(entry.getMessage());
		viewModelLogEntry.setColor(getColorForLogEntry(entry));
		getViewModel().addToLogEntries(viewModelLogEntry);
		logEntryMap.put(entry, viewModelLogEntry);
	}

	protected Color getColorForLogEntry(final LogEntry entry) {
		return Color.BLACK;
	}

	private void removeLogEntry(final LogEntry entry) {
		final var viewModelLogEntry = logEntryMap.remove(entry);
		getViewModel().removeFromLogEntries(viewModelLogEntry);
	}


	/**
	 * Tile Handling
	 */

	private final ListChangeListener<Tile> tilesChangedListener = new ListChangeListener<Tile>() {
		@Override
		public void onChanged(final Change<? extends Tile> change) {
			runLocked(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(tile -> addTileNode(tile));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(tile -> removeTileNode(tile));
					}
				}
			});
		}
	};

	private void addTileNode(final Tile tile) {
		final Location location = tile.getLocation();
		final ListChangeListener<TileContent> contentChangeListener = change -> {
			runLocked(() -> {
				setTileNodeAt(tile.getLocation(), tile);
			});
		};
		tile.contentsProperty().addListener(contentChangeListener);
		tileContentsChangeListeners.put(tile, contentChangeListener);
		setTileNodeAt(location, tile);
	}

	private void removeTileNode(final Tile tile) {
		final var cell = getViewModel().getCellAt(tile.getLocation().getRow(), tile.getLocation().getColumn());
		cell.getLayers().clear();
		removeTileContentsListenerForTile(tile);
	}

	private void removeTileContentsListenerForTile(Tile tile) {
		final ListChangeListener<TileContent> contentChangeListener = tileContentsChangeListeners.remove(tile);
		if (contentChangeListener != null) {
			tile.contentsProperty().removeListener(contentChangeListener);
		}
	}

	private void setTileNodeAt(final Location location, final Tile tile) {
		final var cell = getViewModel().getCellAt(location.getRow(), location.getColumn());
		cell.getLayers().clear();
		onSetTileNodeAtForCell(cell, tile);
	}

	protected abstract void onSetTileNodeAtForCell(final ViewModelCell cell, final Tile tile);

	protected int getRotationForDirection(final Direction direction) {
		assert direction != null;

		switch (direction) {
			case EAST:
				return 0;
			case SOUTH:
				return 90;
			case WEST:
				return 180;
			case NORTH:
				return 270;
		}
		throw new RuntimeException("Invalid direction!");
	}

	@Override
	public void playClicked() {
		miniProgrammingWorld.getPerformance().resumeGame();
	}

	@Override
	public void pauseClicked() {
		miniProgrammingWorld.getPerformance().pauseGame();
	}

	@Override
	public void undoClicked() {
		miniProgrammingWorld.getPerformance().getGameCommandStack().undo();
	}

	@Override
	public void redoClicked() {
		miniProgrammingWorld.getPerformance().getGameCommandStack().redo();
	}

	@Override
	public void speedChanged(final double speedValue) {
		Preconditions.checkState(speedValue >= 0 && speedValue <= 10,
					"Provided speed is not in range [0, 10]");
		miniProgrammingWorld.getPerformance().setSpeed(speedValue);
	}

	@Override
	public void close() {
		miniProgrammingWorld.getPerformance().abortOrStopGame();
	}

	protected void runLocked(final Runnable runnable) {
		try {
			getSemaphore().acquire();
			runnable.run();
		} catch (final Exception e) {
			throw new RuntimeException(e);
		} finally {
			getSemaphore().release();
		}
	}

}

File: core/src/impl/GameViewPresenterBase.cpp
#include "GameViewPresenterBase.h"

#include "Stage.h"
#include "GameLog.h"
#include "MpwSemaphore.h"
#include "CommandStack.h"
#include "GamePerformance.h"

#include "ViewModelLogEntry.h"

#include "CollectionHelpers.hpp"
#include <stdexcept>
#include <utility>

using namespace mpw;
using namespace collectionhelpers;
using namespace framework;

namespace viewmodel {

GameViewPresenterBase::GameViewPresenterBase(std::shared_ptr<mpw::MiniProgrammingWorld> miniProgrammingWorld)
        : inherited(Semaphore()), miniProgrammingWorld(std::move(miniProgrammingWorld)) {
}

void GameViewPresenterBase::bind() {
    auto lock = getSemaphore().lock();

    bindSize();
    bindTiles();
    bindGameLog();
    bindButtons();
    onBind();
}

void GameViewPresenterBase::bindSize() {
    const framework::ObservablePrimitiveProperty<mpw::Size>& sizeProperty = getStageSizeFromConcreteStage();
    GameViewPresenter::getViewModel()->init(sizeProperty.get());
    sizeProperty.addListener([this](auto& oldSize, auto& newSize) {
        auto lock = getSemaphore().lock();
        GameViewPresenter::getViewModel()->init(newSize);
    });
}

void GameViewPresenterBase::bindTiles() {
    auto& tilesProperty = getTilesPropertyFromConcreteStage();

    tilesProperty.addOnAddedListener([this](auto& tile) {
        auto lock = getSemaphore().lock();
        addTileNode(tile);
    });
    tilesProperty.addOnRemovedListener([this](auto& tile) {
        auto lock = getSemaphore().lock();
        removeTileNode(tile);
    });
    tilesProperty.forEach([this](auto& tile) { addTileNode(tile); });
}

void GameViewPresenterBase::onBind() {
    // can be overridden by subclasses
}

void GameViewPresenterBase::bindGameLog() {
    auto miniProgrammingWorldLog = miniProgrammingWorld->getGameLog();
    miniProgrammingWorldLog->logEntriesProperty().addOnAddedListener([this](const LogEntry& entry) {
        auto lock = getSemaphore().lock();
        addLogEntry(entry);
    });
    miniProgrammingWorldLog->logEntriesProperty().addOnRemovedListener([this](const LogEntry& entry) {
        auto lock = getSemaphore().lock();
        removeLogEntry(entry);
    });
    miniProgrammingWorldLog->logEntriesProperty().forEach([this](const LogEntry& entry) {
        addLogEntry(entry);
    });
}

void GameViewPresenterBase::bindButtons() {
    auto viewModel = getViewModel();
    GamePerformance& performance = *miniProgrammingWorld->getPerformance();
    auto& modeProperty = performance.modeProperty();
    auto gameCommandStack = performance.getGameCommandStack();
    modeProperty.addListener([&](Mode, Mode) { updateButtonEnables(); });
    gameCommandStack->executedCommandsProperty().addOnAnyChangeListener(
            [&](const commands::Command&) { updateButtonEnables(); });
    gameCommandStack->undoneCommandsProperty().addOnAnyChangeListener(
            [&](const commands::Command&) { updateButtonEnables(); });
    updateButtonEnables();
    viewModel->setSpeed(performance.getSpeed());
}

void GameViewPresenterBase::updateButtonEnables() {
    auto viewModel = getViewModel();
    auto gameCommandStack = miniProgrammingWorld->getPerformance()->getGameCommandStack();

    Mode mode = miniProgrammingWorld->getPerformance()->getMode();
    bool anyExecutedCommand = !gameCommandStack->getExecutedCommands().empty();
    bool anyUndoneCommand = !gameCommandStack->getUndoneCommands().empty();

    viewModel->setPlayButtonEnabled(mode == Mode::PAUSED);
    viewModel->setPauseButtonEnabled(mode == Mode::RUNNING);
    viewModel->setUndoButtonEnabled(anyExecutedCommand && mode != Mode::RUNNING);
    viewModel->setRedoButtonEnabled(anyUndoneCommand && mode != Mode::RUNNING);
}

void GameViewPresenterBase::playClicked() {
    miniProgrammingWorld->getPerformance()->resumeGame();
}

void GameViewPresenterBase::pauseClicked() {
    miniProgrammingWorld->getPerformance()->pauseGame();
}

void GameViewPresenterBase::undoClicked() {
    miniProgrammingWorld->getPerformance()->getGameCommandStack()->undo();
}

void GameViewPresenterBase::redoClicked() {
    miniProgrammingWorld->getPerformance()->getGameCommandStack()->redo();
}

void GameViewPresenterBase::close() {
    miniProgrammingWorld->getPerformance()->abortOrStopGame();
}

void GameViewPresenterBase::addTileNode(const mpw::Tile& tile) {
    const Location& location = tile.getLocation();

    auto& contentsProperty = tile.contentsProperty();
    addedContentListenerIds[&tile] = contentsProperty.addOnAddedListener([this, &tile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(tile.getLocation(), tile);
    });
    removedContentListenerIds[&tile] = contentsProperty.addOnRemovedListener([this, &tile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(tile.getLocation(), tile);
    });
    setTileNodeAt(location, tile);
}

static void clearListenersForTile(std::unordered_map<const mpw::Tile*, unsigned int>& listenersMap, const mpw::Tile& tile) {
    auto iterator = listenersMap.find(&tile);
    if (iterator != listenersMap.end()) {
        listenersMap.erase(iterator);
    }
}

void GameViewPresenterBase::removeTileNode(const mpw::Tile& tile) {
    auto cell = getViewModel()->getCellAt(
            tile.getLocation().getRow(),
            tile.getLocation().getColumn());
    cell->clearLayers();
    clearListenersForTile(addedContentListenerIds, tile);
    clearListenersForTile(removedContentListenerIds, tile);
}

void GameViewPresenterBase::setTileNodeAt(const Location& location, const Tile& tile) {
    auto cell = getViewModel()->getCellAt(location.getRow(), location.getColumn());
    cell->clearLayers();

    onSetTileNodeAtForCell(*cell, tile);
}

int GameViewPresenterBase::getRotationForDirection(mpw::Direction direction) {
    switch (direction) {
        case Direction::EAST:
            return 0;
        case Direction::SOUTH:
            return 90;
        case Direction::WEST:
            return 180;
        case Direction::NORTH:
            return 270;
    }
    throw std::runtime_error("Invalid direction!");
}

void GameViewPresenterBase::addLogEntry(const LogEntry& entry) {
    auto viewModelEntry = std::make_shared<ViewModelLogEntry>();
    viewModelEntry->setMessage(entry.getMessage());
    viewModelEntry->setColor(getColorForLogEntry(entry));
    getViewModel()->addToLogEntries(viewModelEntry);
    gameLogMapping[&entry] = viewModelEntry.get();
}

void GameViewPresenterBase::removeLogEntry(const LogEntry& entry) {
    auto iterator = gameLogMapping.find(&entry);
    if (iterator != gameLogMapping.end()) {
        ViewModelLogEntry& viewModelEntry = *(*iterator).second;
        getViewModel()->removeFromLogEntries(viewModelEntry.shared_from_this());
        gameLogMapping.erase(iterator);
    }
}

void GameViewPresenterBase::speedChanged(double speedValue) {
    if (!(speedValue >= 0 && speedValue <= 10)) {
        throw std::runtime_error("Provided speed is not in range [0, 10]");
    }
    miniProgrammingWorld->getPerformance()->setSpeed(speedValue);
}

Color GameViewPresenterBase::getColorForLogEntry(const LogEntry& entry) const {
    return Color::BLACK;
}

}
